src/python/pants/backend/adhoc/code_quality_tool.py

```diff
@@ -0,0 +1,549 @@
+    addresses.expect_single()
+
+    tool_targets = await Get(Targets, Addresses, addresses)
+    target = tool_targets[0]
+    runnable_address_str = target[CodeQualityToolRunnableField].value
+    if not runnable_address_str:
+        raise Exception(f"Must supply a value for `runnable` for {request.address}.")
+
+    return CodeQualityTool(
+        runnable_address_str=runnable_address_str,
+        execution_dependencies=target[CodeQualityToolExecutionDependenciesField].value or (),
+        runnable_dependencies=target[CodeQualityToolRunnableDependenciesField].value or (),
+        args=target[CodeQualityToolArgumentsField].value or (),
+        file_glob_include=target[CodeQualityToolFileGlobIncludeField].value or (),
+        file_glob_exclude=target[CodeQualityToolFileGlobExcludeField].value or (),
+        target=target,
+    )
+
+
+@dataclass(frozen=True)
+class CodeQualityToolBatchRunner:
+    digest: Digest
+    args: tuple[str, ...]
+    extra_env: Mapping[str, str]
+    append_only_caches: Mapping[str, str]
+    immutable_input_digests: Mapping[str, Digest]
+
+
+@dataclass(frozen=True)
+class CodeQualityToolBatch:
+    runner: CodeQualityToolBatchRunner
+    sources_snapshot: Snapshot
+    output_files: tuple[str, ...]
+
+
+@rule
+async def process_files(batch: CodeQualityToolBatch) -> FallibleProcessResult:
+    runner = batch.runner
+
+    input_digest = await Get(Digest, MergeDigests((runner.digest, batch.sources_snapshot.digest)))
+
+    result = await Get(
+        FallibleProcessResult,
+        Process(
+            argv=tuple(runner.args + batch.sources_snapshot.files),
+            description="Running code quality tool",
+            input_digest=input_digest,
+            append_only_caches=runner.append_only_caches,
+            immutable_input_digests=FrozenDict.frozen(runner.immutable_input_digests),
+            env=FrozenDict(runner.extra_env),
+            output_files=batch.output_files,
+        ),
+    )
+    return result
+
+
+@rule
+async def hydrate_code_quality_tool(
+    request: CodeQualityToolAddressString,
+) -> CodeQualityToolBatchRunner:
+    cqt = await Get(CodeQualityTool, CodeQualityToolAddressString, request)
+
+    runnable_address = await Get(
+        Address,
+        AddressInput,
+        AddressInput.parse(
+            cqt.runnable_address_str,
+            relative_to=cqt.target.address.spec_path,
+            description_of_origin=f"Runnable target for code quality tool {cqt.target.address.spec_path}",
+        ),
+    )
+
+    addresses = Addresses((runnable_address,))
+    addresses.expect_single()
+
+    runnable_targets = await Get(Targets, Addresses, addresses)
+
+    target = runnable_targets[0]
+
+    field_sets = await Get(
+        FieldSetsPerTarget, FieldSetsPerTargetRequest(RunFieldSet, runnable_targets)
+    )
+
+    environment_name = await Get(
+        EnvironmentName, EnvironmentNameRequest, EnvironmentNameRequest.from_target(target)
+    )
```

Comment:
It'd be good to run these requests concurrently as much as possible with `MultiGet` https://www.pantsbuild.org/docs/rules-api-concepts#multiget-for-concurrency, for instance: 

```suggestion
    field_sets, environment_name = await MultiGet(
        Get(
            FieldSetsPerTarget, FieldSetsPerTargetRequest(RunFieldSet, runnable_targets)
        ),
        Get(
            EnvironmentName, EnvironmentNameRequest, EnvironmentNameRequest.from_target(target)
        )
    )
```

Potentially the `execution_environment` request can be included in this `MultiGet` too.

I think maybe the `merged_extras` and `digest` `Get`s could be combined into a `MultiGet` too.

So, the critical path through this code might be something like

```mermaid
flowchart TD
  cqt --> runnable_address
  subgraph MultiGet1
  field_sets
  environment_name
  execution_environment
  end
  runnable_address --> MultiGet1
  MultiGet1 --> run_request
  subgraph MultiGet2
  merged_extras
  digest
  end
  run_request --> MultiGet2
```

In general, preferably any `Get`s that are independent run concurrently via a `MultiGet`. It looks like there maybe some others in some other rules too.

Reply:
And I did not know you can make diagrams like that in github. Thank you!